### PR TITLE
Curl to install directpv

### PR DESCRIPTION
**Pull request by sinhaashish**
Run the below command
```
curl -s https://raw.githubusercontent.com/minio/directpv/master/install-script.sh | bash
```

### CODE REVIEW

File: install-script.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (balamurugana) [inline install-script.sh:1]: What is the use to wrap just `kubectl directpv install` command?

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: install-script.sh [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+echo "Installing direct PV"
+
+if [[ $kv = 0 ]]; then
```
Comment (balamurugana) [inline install-script.sh:21]: This is never going to work. You should execute `kubectl version` command and check if it succeeds.

File: install-script.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    echo "kubectl found"
+else
+    curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
+    chmod +x kubectl
+    mkdir -p ~/.local/bin/kubectl
+    mv ./kubectl ~/.local/bin/kubectl
```
Comment (balamurugana) [inline install-script.sh:27]: Download directly into `~/.local/bin` directory

File: install-script.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+echo "Installing direct PV"
+
+if [[ $kv = 0 ]]; then
+    echo "kubectl found"
```
Comment (balamurugana) [inline install-script.sh:22]: No need to show this.

File: install-script.sh [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+if [[ $krew-version = 0 ]]; then
+    echo "krew found"
+else
+    set -x; cd "$(mktemp -d)" &&
+    curl -fsSLO "https://github.com/kubernetes-sigs/krew/releases/latest/download/krew.tar.gz" &&
+    tar zxvf krew.tar.gz &&
+    KREW=./krew-"$(uname | tr '[:upper:]' '[:lower:]')_$(uname -m | sed -e 's/x86_64/amd64/' -e 's/arm.*$/arm/')" &&
+    "$KREW" install krew
+    export PATH="${KREW_ROOT:-$HOME/.krew}/bin:$PATH"
+    echo "krew installed"
```
Comment (balamurugana) [inline install-script.sh:40]: Same as above

File: install-script.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+fi
+# Install kubectl directpv plugin
+if [[ $krew = 0 ]]; then
+    echo "Installed kubectl directpv plugin"
+else
+    echo "Problem in installing krew DirectPV plugin"
+fi
+
```
Comment (balamurugana) [inline install-script.sh:48]: You have to install `directpv` plugin first and install directpv in k8s.

File: install-script.sh [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    operating_system=$(fetch_OS)
+    verify_kubectl $operating_system
+    verify_krew $operating_system
+    kubectl krew install directpv
+    kubectl directpv install
+}
+
+# Fetch the operating system
+function fetch_OS(){
+    if [ "$(uname)" == "Darwin" ]; then
```
Comment (balamurugana) [inline install-script.sh:37]: use `case` instead of nested `if` everywhere

File: install-script.sh [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    fi
+    echo "$operating_system"
+}
+
+
+function verify_kubectl() {
+    # check for kubectl installation
+    kubectl version --client
+    # Install kubectl if not present
+    if [ $? -ne 0 ]; then
```
Comment (balamurugana) [inline install-script.sh:54]: use `if COMMAND; then` everywhere

File: install-script.sh [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        set PATH=%PATH%;$PWD
+    fi
+}
+
+function verify_minikube() {
+    minikube version
+    if [ $? -ne 0 ]; then
+        install_minikube "$1"
+    fi
+    minikube start
```
Comment (balamurugana) [inline install-script.sh:87]: why do we need `minikube`?

File: install-script.sh [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+function install_krew() {
+    git --version
+    if [ $? -ne 0 ]; then
+        echo "Please install git and try again."
+    fi
+
+    if [ "$1" == $darwin ] || [ "$1" == $linux ]; then
+        set -x; cd "$(mktemp -d)" &&
```
Comment (balamurugana) [inline install-script.sh:118]: use `set -e` instead of `&&`

File: install-script.sh [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/darwin/amd64/kubectl"
+        chmod +x kubectl
+        cd -
+    elif [ "$1" == $linux ]; then
+        cd ~/.local bin
+        curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
+        chmod +x kubectl
+        cd -
+    elif [ "$1" == $windows ]; then
+        curl -LO "https://dl.k8s.io/release/v1.23.0/bin/windows/amd64/kubectl.exe"
```
Comment (balamurugana) [inline install-script.sh:77]: have the url constructed at the tme of finding OS

File: install-script.sh (hunk 9 above)
Comment (sinhaashish) [reply install-script.sh:89]: If kubectl client is there , but its not configured , then this PR install minikube and runs kubectl against it

File: functests/install-script.sh [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+            chmod +x kubectl 
+            cd - ;;
+        linux)
+            cd ~/.local/bin
+            curl -LO $kubectl_url 
+            chmod +x kubectl 
+            cd - ;;
+        windows)
+            curl -LO $kubectl_url
+            PATH=%PATH%;$PWD ;;
```
Comment (balamurugana) [inline functests/install-script.sh:81]: `Warning: ./functests/install-script.sh:81:13: warning: PATH is the shell search path. Use another name. [SC2123]`

You cannot mix MS Windows PATH variable and bash PATH variable notation.

File: functests/install-script.sh [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+echo "Installing direct PV"
+
+operating_system=""
+darwin="Darwin"
+linux="Linux"
+windows="Windows"
+
+function directpv(){
+    read operating_system kubectl_url minikube_url < <(initialize)
```
Comment (balamurugana) [inline functests/install-script.sh:27]: export these variables in `initialize` function.

File: functests/install-script.sh [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    verify_kubectl
+    verify_krew 
+    kubectl krew install directpv
+    kubectl directpv install
+}
+
+# Fetch the operating system
+function initialize(){
```
Comment (balamurugana) [inline functests/install-script.sh:35]: just rename this function to `init`

File: functests/install-script.sh [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+            operating_system=$darwin            
+            kubectl_url="https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/darwin/amd64/kubectl"
+            minikube_url="https://storage.googleapis.com/minikube/releases/latest/minikube-darwin-amd64"
+            ;;
+        linux*)   
+            operating_system=$linux
+            kubectl_url="https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
+            minikube_url="https://storage.googleapis.com/minikube/releases/latest/minikube-linux-amd64"
+            ;;
+        msys*)
```
Comment (balamurugana) [inline functests/install-script.sh:47]: Are you working with `msys2` on MS Windows or are we supporting https://itsfoss.com/install-bash-on-windows/?

File: functests/install-script.sh [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+darwin="Darwin"
+linux="Linux"
+windows="Windows"
+
+function directpv(){
+    read operating_system kubectl_url minikube_url < <(initialize)
+    verify_kubectl
+    verify_krew 
+    kubectl krew install directpv
+    kubectl directpv install
```
Comment (balamurugana) [inline functests/install-script.sh:31]: To install `directpv`, we need `KUBECONFIG` set by the user to install on decided k8s for the user.

File: functests/install-script.sh (hunk 12 above)
Comment (sinhaashish) [reply functests/install-script.sh]: This PR caters Linux only .

**Pull request closed without merging**